### PR TITLE
[AIRFLOW-5266] Allow aws_athena_hook to get all query results

### PR DESCRIPTION
**Pull request by lindsable**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-5266#Athena.Client.get_query_results%5D) issues and references them in the PR title. 

### Description

- [x] The previous version of the hook returned up to 1000 results. The update allows the user to set the max results (default and max constrained by the api is 1000) and also the NextTokenId.  
- [x] The client request token in the run_query function is now truly optional. Boto3 doesn't allow you to send in parameters that are set to None.
- [x] Another function was added to return a paginator for the query results. 

### Tests

- [x] My PR adds the following unit tests:
tests/providers/amazon/aws/hooks/test_athena.py 

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by lindsable** [comment]
The ci build failed because pylint didn't accept the boto3 parameters I monkey patched. Should I add an ignore to the test file for those?

File: airflow/contrib/hooks/aws_athena_hook.py [hunk 1]
```diff
@@ -157,6 +178,7 @@ def poll_query_status(self, query_execution_id, max_tries=None):
                 break
             if max_tries and try_number >= max_tries:  # Break loop if max_tries reached
                 final_query_state = query_state
+                self.stop_query(query_execution_id)
```
Comment (feluelle) [inline airflow/contrib/hooks/aws_athena_hook.py]: You should mention that in the commit :)

It fixes a bug, right?

File: tests/contrib/hooks/test_aws_athena_hook.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import unittest
+from unittest import mock
+
+try:
+    from airflow.contrib.hooks.aws_athena_hook import AWSAthenaHook
+except ImportError:
+    AWSAthenaHook = None  # type: ignore
+
+
+class MockAthenaClient:
```
Comment (feluelle) [inline tests/contrib/hooks/test_aws_athena_hook.py]: I am not a fan of this, because you have basically two `AthenaHook`s you need to maintain.
If you use `mock.patch` you can achieve the same without having two classes to keep in sync.

Comment (mik-laj) [reply tests/contrib/hooks/test_aws_athena_hook.py]: I agree. We should use `unittest.mock` in this place.

File: airflow/contrib/hooks/aws_athena_hook.py (hunk 1 above)
Comment (lindsable) [reply airflow/contrib/hooks/aws_athena_hook.py]: Yup!

File: tests/contrib/hooks/test_aws_athena_hook.py (hunk 2 above)
Comment (lindsable) [reply tests/contrib/hooks/test_aws_athena_hook.py]: I had an issue mocking this function because of the way boto3 implements the api. It's not a defined function but an api call. There are a few suggestions that I found [here:](https://stackoverflow.com/questions/37143597/mocking-boto3-s3-client-method-python) but we would still have to maintain it. The other aws hooks that use boto3 are able to use the [moto](https://github.com/spulec/moto) package.  but athena isn't mocked yet

Comment (OmerJog) [reply tests/contrib/hooks/test_aws_athena_hook.py]: @mik-laj @feluelle  is there a solution for this ? Did some checks everyone everywhere complains it's impossible to mock athena with moto. How shall we proceeded?

Comment (feluelle) [reply tests/contrib/hooks/test_aws_athena_hook.py]: You don't need the moto package to mock a class. Just mock the calls that are accessing the boto library in the get_query_results function. So that it will never really call the content of the moto get_query_results function.

Comment (feluelle) [reply tests/contrib/hooks/test_aws_athena_hook.py]: Actually you only need to add `mock_conn.get_query_results.return_value = self.base_query_result` to the beginning of the test function. That is basically what you are already doing.

You are doing good actually. But I would remove the extra class.
This part
```
        if NextToken is None:
            return self.query_result_with_next
        else:
            return self.base_query_result
```
is logic that you don't need if you set the `mock_conn.get_query_results.return_value` directly in the test function.

A MockClass should not contain extra logic.

Do you understand what I mean? :D

Comment (lindsable) [reply tests/contrib/hooks/test_aws_athena_hook.py]: Fixed!

File: airflow/providers/amazon/aws/hooks/athena.py [hunk 3]
```diff
@@ -133,7 +133,17 @@ def get_query_results(self, query_execution_id):
         elif query_state in self.INTERMEDIATE_STATES or query_state in self.FAILURE_STATES:
             self.log.error('Query is in {state} state. Cannot fetch results'.format(state=query_state))
             return None
-        return self.get_conn().get_query_results(QueryExecutionId=query_execution_id)
+        result_params = {'QueryExecutionId': query_execution_id,
+                         'MaxResults': max_results}
+        results = self.get_conn().get_query_results(**result_params)
```
Comment (ashb) [inline airflow/providers/amazon/aws/hooks/athena.py]: Does boto have a built-in pager for this?

Comment (ashb) [reply airflow/providers/amazon/aws/hooks/athena.py]: Yes, see https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/athena.html#paginators and https://boto3.amazonaws.com/v1/documentation/api/latest/guide/paginators.html

File: tests/providers/amazon/aws/hooks/test_athena.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import unittest
+from unittest import mock
+
+try:
+    from ..airflow.providers.amazon.aws.hooks.athena import AWSAthenaHook
```
Comment (ashb) [inline tests/providers/amazon/aws/hooks/test_athena.py]: `..` here is odd/unusual, please just use absolute import here

File: tests/providers/amazon/aws/hooks/test_athena.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
```
Comment (ashb) [inline tests/providers/amazon/aws/hooks/test_athena.py]: Do we really not have any existing tests for the Athena hook? (i.e. Why is this a new file?)

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/providers/amazon/aws/hooks/athena.py (hunk 3 above)
Comment (lindsable) [reply airflow/providers/amazon/aws/hooks/athena.py]: Yes that does make sense.  Now that I think about it, it probably makes sense to just add a new function to check that the query is successful and then return the paginator. For the get_query_results function maybe we should just adding the option to send in the NextTokenId so you can paginate manually if you want.

File: tests/providers/amazon/aws/hooks/test_athena.py (hunk 5 above)
Comment (lindsable) [reply tests/providers/amazon/aws/hooks/test_athena.py]: That is correct.

File: tests/providers/amazon/aws/hooks/test_athena.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                }
+            ]
+        }
+    },
+    'NextToken': 'MockNextToken'
+}
+MOCK_GET_QUERY_EXECUTION = {'QueryExecution': {'Status': {'State': 'SUCCEEDED'}}}
+MOCK_ATHENA_QUERY_ID = 'MockQueryExecutionId'
+
+# All other functions are tested in test/contrib/operators/test_aws_athena_operator.py
```
Comment (feluelle) [inline tests/providers/amazon/aws/hooks/test_athena.py]:
```suggestion
```

File: airflow/providers/amazon/aws/hooks/athena.py [hunk 7]
```diff
@@ -116,7 +120,7 @@ def get_state_change_reason(self, query_execution_id):
             # The error is being absorbed to implement retries.
             return reason  # pylint: disable=lost-exception
 
-    def get_query_results(self, query_execution_id):
+    def get_query_results(self, query_execution_id, next_token_id=None, max_results=1000):
```
Comment (feluelle) [inline airflow/providers/amazon/aws/hooks/athena.py:116]: Please add docs regarding next_token_id and max_results

File: airflow/providers/amazon/aws/hooks/athena.py [hunk 8]
```diff
@@ -132,7 +136,37 @@ def get_query_results(self, query_execution_id):
...
-        return self.get_conn().get_query_results(QueryExecutionId=query_execution_id)
+        result_params = {
+            'QueryExecutionId': query_execution_id,
+            'MaxResults': max_results
+        }
+        if next_token_id:
+            result_params['NextToken'] = next_token_id
+        return self.get_conn().get_query_results(**result_params)
+
+    def get_query_results_paginator(self, query_execution_id, max_results=1000):
```
Comment (feluelle) [inline airflow/providers/amazon/aws/hooks/athena.py]: Same here.

**Review by feluelle — APPROVED** [review]

**Comment by lindsable** [comment]
Let me know if you want me to squash any of my commits or rebase with master

**Comment by feluelle** [comment]
I think it is fine.

If @ashb is also fine with it, he can `Squash and merge` it. He requested changes so I would like him to do the last review. :)

**Comment by ashb** [comment]
I'll take a look at this tomorrow

**Comment by dephusluke** [comment]
I took the core from it and created my own custom hook and it seems to work.
Once this will be release I'll switch to the official hook

File: airflow/providers/amazon/aws/hooks/athena.py [hunk 9]
```diff
@@ -163,6 +213,7 @@ def poll_query_status(self, query_execution_id, max_tries=None):
                 break
             if max_tries and try_number >= max_tries:  # Break loop if max_tries reached
                 final_query_state = query_state
+                self.stop_query(query_execution_id)
```
Comment (ashb) [inline airflow/providers/amazon/aws/hooks/athena.py]: This line appears to be unrelated to fetching all results or not, right?

I guess this is probably a sane default, but this is a change in the current behaviour. Previously any query would continue running meaning that even if it takes a lot longer in Airflow it would (eventually) be available in the Athena console. 

Hmmm.

Comment (ashb) [reply airflow/providers/amazon/aws/hooks/athena.py]: Other than this (which I don't know if we want or not) the code looks good, and tests are always great :)

**Review by ashb — COMMENTED** [review]
Sorry it took rather more than the promised "tomorrow".

File: airflow/providers/amazon/aws/hooks/athena.py (hunk 9 above)
Comment (ashb) [reply airflow/providers/amazon/aws/hooks/athena.py]: @lindsable What do you think? Is this required? Is it a sane default? Are there cases where users might want the query to continue anyway?

Comment (lindsable) [reply airflow/providers/amazon/aws/hooks/athena.py]: While I can't think of anything specific, I think you have a good point about changing existing functionality. We will leave it up to the end user to stop query if they choose. I'll remove this line.

Comment (ashb) [reply airflow/providers/amazon/aws/hooks/athena.py]: Can you ping me once this has been changed?

(Sorry this ticket has sat for _so_ long.)

Comment (lindsable) [reply airflow/providers/amazon/aws/hooks/athena.py]: Changed! Just waiting for the tests to pass again. @ashb

Comment (lindsable) [reply airflow/providers/amazon/aws/hooks/athena.py]: @ashb I'm not sure why the static checks are failing on the docker lint. Any insights?

Comment (ashb) [reply airflow/providers/amazon/aws/hooks/athena.py]: @lindsable Yeah does seem unrelated to your change -- try rebasing on to latest master.

Comment (lindsable) [reply airflow/providers/amazon/aws/hooks/athena.py]: Looks like the same tests are also failing on master :(

**Comment by mik-laj** [comment]
@lindsable Do you need any help?

If you have applied all the comments of the reviewer, then it is worth asking to review again using a comment or PM on Slack. The commiters have a lot of work and are unable to check the status of every change.

   > Ping @ #development slack, comment @people. Be annoying. Be considerate.

   https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#id33

**Comment by lindsable** [comment]
Sorry for the delay but I was able to rebase today and get all of the tests passing! @ashb Looks like everything is ready to merge!

**Comment by lindsable** [comment]
@ashb I'm sure things have been busy for you organizing the summit, but I'd really like to get this merged. Please let me know what I can do.

File: airflow/providers/amazon/aws/hooks/athena.py [hunk 10]
```diff
@@ -125,7 +133,49 @@ def get_query_results(self, query_execution_id):
...
+        :type max_items: int
+        :param page_size: The size of each page.
+        :type page_size: int
+        :param starting_token: A token to specify where to start paginating.
+        :type starting_token: str
+        :return: PageIterator
+        """
+        query_state = self.check_query_status(query_execution_id)
+        if query_state is None:
+            self.log.error('Invalid Query state')
```
Comment (ashb) [inline airflow/providers/amazon/aws/hooks/athena.py]:
```suggestion
            self.log.error('Invalid Query state (null)')
```

File: airflow/providers/amazon/aws/hooks/athena.py [hunk 11]
```diff
@@ -125,7 +133,49 @@ def get_query_results(self, query_execution_id):
...
+            return None
+        elif query_state in self.INTERMEDIATE_STATES or query_state in self.FAILURE_STATES:
```
Comment (ashb) [inline airflow/providers/amazon/aws/hooks/athena.py]:
```suggestion
        if query_state in self.INTERMEDIATE_STATES or query_state in self.FAILURE_STATES:
```

File: tests/providers/amazon/aws/hooks/test_athena.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+MOCK_RUNNING_QUERY_EXECUTION = {'QueryExecution': {'Status': {'State': 'RUNNING'}}}
+MOCK_SUCCEEDED_QUERY_EXECUTION = {'QueryExecution': {'Status': {'State': 'SUCCEEDED'}}}
+
+MOCK_QUERY_EXECUTION = {'QueryExecutionId': MOCK_DATA['query_execution_id']}
+
+
+class TestAWSAthenaHook(unittest.TestCase):
+
+    def setUp(self):
+        self.athena = AWSAthenaHook(sleep_time=1)
```
Comment (ashb) [inline tests/providers/amazon/aws/hooks/test_athena.py]: Can we do sleep time=0 instead? (`sleep(0)` is valid) - This would probably make the tests run quicker.

**Review by ashb — CHANGES_REQUESTED** [review]
A few tiny changes, this is looking good.

Sorry it's been open for so very long!

**Review by ashb — APPROVED** [review]
